### ISSUE TITLE
Allow multiple hooks given by cli

### ISSUE THREAD

**Issue by Zocker1999NET** [certbot/certbot#8090]
## My operating system is (include version):
- Debian GNU/Linux Buster (currently Stable)
- Certbot version: 0.31.0

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):
- OS package manager

## I ran this command:
```sh
certbot certonly --non-interactive --disable-hook-validation --post-hook "systemctl reload nginx" --post-hook "systemctl restart nginx" --domain example.com
```
Notes:
- Webroot mode is configured using `cli.ini`, but this should not affect this issue / feature request
- The similar hook commands are for testing and illustration purposes

The result was that nginx was only restarted, not reloaded.

## Certbot's behavior differed from what I expected because:
I expected nginx first to be reloaded, then to be restarted.

This is not an issue with systemd because systemd allows services to be reloaded and restarted in these short time periods, otherwise it will log a failure if a service was not able to handle that, which does not happened in my case.

Hereby I suggest to add the support of multiple reload commands. Certbot seems to support to recognize identical reload commands so they will not be executed multiple times (see https://certbot.eff.org/docs/using.html#renewing-certificates). So it would be great, if I could configure each reload command individually so Certbot does reload the service only once and not five or more times (currently on my server the case every 60 days). Because some certificates require additional reload commands and I needed to add them using `&&`, Certbot is not able to recognize identical commands.

**Comment by bmw** [comment]
Hi @Zocker1999NET. Thanks for the detailed issue.

A design like you proposed seems complicated to me because it adds a command ordering that probably should be respected in addition to the logic we already have that tries to skip running the same command multiple times on a run. I think respecting this ordering while running each command once adds a lot of unnecessary complexity to Certbot.

Instead, for your certificates that need to reload/restart nginx, what do you think about using `systemctl reload nginx && systemctl restart nginx` for all of them? If they usually all renew at the same time like you say, I think that should have the effect you want.

**Comment by Zocker1999NET** [comment]
Thank you for your answer, but this does not help me. I've missed to state some things because I tried to make the example above easy. My use case involves following problems:
- I want to use certbot to track each certificates individual reload commands, as some of my certificates require special reload commands which are not required for the other certs.
- Each certificate got its own "personal" reload command, some applications require a combined privkey and cert file (see #5087), so each certificate also has a reload command `cat <PRIVKEY>.pem <CERTFILE>.pem > <COMBINED>.pem` and this command is different for each certificate because of the paths. So if each certificate will be configured with the reload command `cat … && systemctl reload nginx`, nginx will be reloaded once per certificate due to certbot not recognizing this command.

I think you can easily respect the order of commands by collecting all reload commands per certificate in the order they were given by the user, combine all together to a single list (order of certificates should not matter), and then remove duplicates from end to the beginning. So each command will be called once and the user and every command required should be already executed.

Example: Two certificates, namely example.com and website.com, were renewed. The have following reload commands:
```yaml
example.com:
  - cat example.com/privkey.pem example.com/fullchain.pem > example.com/combined.pem
  - systemctl reload nginx
website.com:
  - cat website.com/privkey.pem website.com/fullchain.pem > website.com/combined.pem
  - systemctl reload nginx
  - systemctl reload postfix # Required only for website.com
```
It is obvious that in this case, nginx and postfix need to be reloaded **after** the combined file were created (assuming the services require those). The combined list will be:
```yaml
  - cat example.com/privkey.pem example.com/fullchain.pem > example.com/combined.pem
  - systemctl reload nginx # Duplicate
  - cat website.com/privkey.pem website.com/fullchain.pem > website.com/combined.pem
  - systemctl reload nginx # Duplicate
  - systemctl reload postfix
```
And removing duplicate commands from end to beginning:
```yaml
  - cat example.com/privkey.pem example.com/fullchain.pem > example.com/combined.pem
  - cat website.com/privkey.pem website.com/fullchain.pem > website.com/combined.pem
  - systemctl reload nginx # Note this one is below both cat commands
  - systemctl reload postfix
```
I think this fulfills all expectations of the users giving multiple reload commands.

I do not have any other idea how to implement this using certbot without modifiying it, and given the algorithm above does not feel very complicated, I would implement this into certbot if the change will not be rejected.

**Comment by bmw** [comment]
I do not think it is trivial to respect the order of commands because of the possibility of conflicting ordering. For a trivial example, if the user had
```
example.com:
  - systemctl reload postfix
  - systemctl reload nginx
website.com:
  - systemctl reload nginx
  - systemctl reload postfix
```
it is impossible to both run each command once and run them in the order specified by the user for each certificate. In this case, I think a user could pretty easily set up this configuration themselves if they set up the hooks at different times and the order doesn't matter. You could also have much more complicated examples when more certificates/commands are involved.

Because of this and the potential of problems when order does matter such as your examples  with `cat`, I don't think we're going to accept this change so I'm going to close this issue as wontfix.

If you would like to continue to optimize your Certbot setup to avoid unnecessary commands, I would encourage you to look at:

1. The directories under `/etc/letsencrypt/renewal-hooks` which are documented a bit under https://certbot.eff.org/docs/using.html#renewing-certificates. These directories allow you to create executables that define common hooks for all certificates and are run in addition to hooks specified for each certificate. Executables under `/etc/letsencrypt/renewal-hooks/pre` and `/etc/letsencrypt/renewal-hooks/post` will only be run once per `certbot renew` run.
2. Deploy hooks which run before post hooks and can be defined with `--deploy-hook` or under `/etc/letsencrypt/renewal-hooks/deploy`. These hooks are run after renewing the certificate and extra environment variables are set providing information about the certificate that was renewed which may allow you to generalize some of your certificate specific commands. The help output for the flag is:
```
  --deploy-hook DEPLOY_HOOK
                        Command to be run in a shell once for each
                        successfully issued certificate. For this command, the
                        shell variable $RENEWED_LINEAGE will point to the
                        config live subdirectory (for example,
                        "/etc/letsencrypt/live/example.com") containing the
                        new certificates and keys; the shell variable
                        $RENEWED_DOMAINS will contain a space-delimited list
                        of renewed certificate domains (for example,
                        "example.com www.example.com" (default: None)
```
I hope the combination of these things helps.
